Add restart policy handling for workloads (#249)
eclipse-ankaios/ankaios: agent/doc/swdesign/README.md

### Restart policies of workloads

#### Agent supports restart policies

> `swdd~agent-supports-restart-policies~2`

Author: Reverted.

- impl
- utest

#### WorkloadControlLoop handles restarts of workloads
`swdd~workload-control-loop-handles-workload-restarts~1`

Status: approved

When the WorkloadControlLoop receives a new workload state, then the WorkloadControlLoop shall:
- triggers the check of the permissibility of the restart

> - execute the restart of the workload it manages based on the result of the comparison

Author: The unit test would be much more complicated because to test the if inside the select code I must test the run method which is complicated to test because of the tokio setup code and timeout. This way the function covers the restart handling and I test the private function which is easier.

> - execute the restart of the workload it manages based on the result of the comparison

Author: I have moved the first action to a precondition.

- utest

#### WorkloadControlLoop restarts workload with enabled restart policy
`swdd~workload-control-loop-restarts-workload-with-enabled-restart-policy~1`

Status: approved

When:
the workload's `ExecutionState` is `Succeeded(Ok)` or `ExecutionState` is `Failed(ExecFailed)` and the workload's configured `RestartPolicy` contains the value `ALWAYS` or
the workload's `ExecutionState` is `Failed(ExecFailed)` and the workload's configured `RestartPolicy` contains the value `ON_FAILURE`,
then the WorkloadControlLoop shall restart the workload.

Rationale:
The restart depends on the execution state of the workload.

Tags:
- WorkloadControlLoop

Needs:
- impl
- utest
- stest

#### WorkloadControlLoop does not restart a workload with disabled restart policy
`swdd~workload-control-loop-no-restart-with-disabled-restart-policy~1`

Status: approved

When the workload's configured `RestartPolicy` contains the value `NEVER`, then the WorkloadControlLoop shall not restart the workload.

Rationale:
The user has explicitly configured the workload to prevent its restart or to apply the default value of "NEVER" in the workload configuration.

Tags:
- WorkloadControlLoop

Needs:
- impl
- utest

> - stest

Reviewer: Do we really need this requirement as a separate one?
Maybe we can merge it with the one above by just stating in a comment that the "NEVER" policy is there for completeness and is handled as policy is never met.
(Normally one does not write negative requirements.)

Author: I can write it into a comment, but I thought it is more expressive to link it to the unit test explicitly as an extra requirement. I thought this is an exception here, because it is important that a workload is not restarted with policy `NEVER` and I did not want to cover this in a "weak" comment.

Author: I deleted this negative req and linked the one to the stest where inside the comment is now mentioned that with `NEVER` no restart happens.

#### WorkloadControlLoop restarts workloads using the update operation
`swdd~workload-control-loop-restarts-workloads-using-update~1`

Status: approved

When the WorkloadControlLoop executes restarts of the workload, then the WorkloadControlLoop shall:
- delete the existing workload via the corresponding runtime connector
- create a new workload with the stored workload configuration via the corresponding runtime connector

Comment:
The restart is represented within the system by an update operation.

Rationale:
A runtime may not support directly restarting the exited container.

Tags:
- WorkloadControlLoop

Needs:
- impl
- utest

#### WorkloadControlLoop skips restarts
`swdd~workload-control-loop-skips-restarts~1`

> When the WorkloadControlLoop handles a restart of a workload and the workload's current `WorkloadInstanceName` inside the WorkloadControlLoop differs from the `WorkloadInstanceName` inside the received workload state, then the WorkloadControlLoop shall skip the restart.

Reviewer: We have 3 requirements about the preconditions of the restart, but they are somehow not really connected together.
Maybe we should mention this here in the requirement above and reorder a bit:

1. this requirement here with a comment like "If the workload changed in the meantime, there is no need to restart it."
2. one requirement about the execution states which states "If the worklaod has not changed and either the workload's `ExecutionS.... or .... the workload's `ExecutionState` is `Failed ... then"

Author: I have fixed it.

- impl
- utest

#### WorkloadControlLoop handles restarts of workloads
`swdd~workload-control-loop-handles-workload-restarts~1`

Status: approved

> When the WorkloadControlLoop receives a new workload state and the WorkloadControlLoop checks if the restart of the workload it manages is required, then the WorkloadControlLoop shall execute the restart of the workload it manages based on the result of the comparison

Reviewer: ```suggestion
When the WorkloadControlLoop receives a new workload state for a workload it manages and the WorkloadControlLoop detects that a restart of the workload is required according to its restart policy, then the WorkloadControlLoop shall execute the restart of the workload.
```

Rationale:
The execution of a restart of the workload depends on the workload state and the configured restart policy.

Tags:
- WorkloadControlLoop

Needs:
- impl
- utest

#### WorkloadControlLoop skips restarts
`swdd~workload-control-loop-skips-restarts~1`

When the WorkloadControlLoop handles a restart of a workload and the workload's current `WorkloadInstanceName` inside the WorkloadControlLoop differs from the `WorkloadInstanceName` inside the received workload state, then the WorkloadControlLoop shall skip the restart.

If the workload changed in the meantime, there is no need to restart it.

Rationale:

> This prevents wrong restarts of workloads when a new workload state of the old workload is received during an update operation runs.

Reviewer: ```suggestion
This prevents unneeded restarts of workloads when a new workload state of the old workload is received after an update operation was performed.
```

Tags:
- WorkloadControlLoop

Needs:
- impl
- utest

#### WorkloadControlLoop restarts workload with enabled restart policy
`swdd~workload-control-loop-restarts-workload-with-enabled-restart-policy~1`

Status: approved

When the workload has not changed and:
* either the workload's `ExecutionState` is `Succeeded(Ok)` or `ExecutionState` is `Failed(ExecFailed)` and the workload's configured `RestartPolicy` contains the value `ALWAYS`
* or the workload's `ExecutionState` is `Failed(ExecFailed)` and the workload's configured `RestartPolicy` contains the value `ON_FAILURE`,
then the WorkloadControlLoop shall restart the workload.

Comment:

> If the workload changed in the meantime, there is no need to restart it. In case of the restart policy is `NEVER` the workload is not restarted.

Reviewer: The first sentence is clarifying a different requirement, we can remove it.
```suggestion
In all other execution states and in case of the restart policy is `NEVER` the workload is not restarted.
```
